### DISCUSSION TITLE
SaveAsync behavior

### DISCUSSION THREAD

**Discussion by CROprogrammer** [MarimerLLC/csla#3283] [Questions]
I've created my business object in which I have implemented [Create] and [Insert] methods, and I'm using Blazor for UI. I didn't want that business rules are checked on creation of business object because then I get validation messages even though inputs have not been touched. Then I tried to remove BusinessRules.CheckRules() from Create method to not get those messages but then I'm able to save empty object in database because my business rules are not checked when I click save button that calls ViewModel.SaveAsync() method. Why is that. How can I accomplish that business rules are checked after I click save button and that empty object is not going to be saved but business rules are going to be checked beforehand.
I've tried to put BusinessRules.CheckRules() in the body of Insert method but then it works just the first time I try to save my object (it checks the business rules and I get validation messages) but then after when I insert valid data in my form and click save nothing happens (SaveAsync does not even call Insert method again).

TLDR: How can I accomplish that SaveAsync method in Blazor checks business rules before calling Insert method in BusinessBase object

Im using CSLA 5.5.3

**Comment by rockfordlhotka** [comment]
Have you looked at overriding SaveAsync in the business class? You should be able to do something like this pseudocode:

```c#
  protected override SaveAsync()
  {
    if (this.IsNew)
    { 
      await BusinessRules.CheckRulesAsync();
    }
    return await base.SaveAsync();
  }
```

**Reply by CROprogrammer** [reply]
I tried that - not with the overriding SaveAsync though because that method cannot be overriden, but with hidding base method. After that I binded that method to my save button (@onclick="vm.Model.SaveAsync") and it works initially when I try to save invalid model, and validation messages are shown. But then after that when I put valid data in my form validation messages don't go away and nothing happens on button click. Any idea why?

**Reply by CROprogrammer** [reply]
That is my problem now when I try to call ViewModel.Model.SaveAsync (SaveAsync method from class that extends BusinessBase) on click.

When I call CSLA Blazor's ViewModel.SaveAsync that's when I have a problem described in the first post.

Is there any way to override CSLA Blazor's ViewModel.SaveAsync method so that it first checks business rules and if they are broken then do not call [Insert] method from business base class. I supposed that that would be expected behaviour.

**Reply by rockfordlhotka** [reply]
There was an oversight for a while, that `BusinessBase.SaveAsync` wasn't virtual. That was fixed, but I don't remember when.

That's the method you need to override though. Not anything in the viewmodel.
